lib/reversetunnel/track/tracker.go

```diff
@@ -52,268 +45,359 @@ func (c *Config) CheckAndSetDefaults() error {
-}
-
-// TrackExpected starts/refreshes tracking for expected proxies.  Called by
-// agents when gossip messages are received.
-func (t *Tracker) TrackExpected(proxies ...string) {
+// canSpawn returns true if the current state of the Tracker is such that the
+// client should attempt to spawn new connections.
+func (t *Tracker) canSpawn() bool {
 	t.mu.Lock()
 	defer t.mu.Unlock()
 
-	if t.sets == nil {
-		return
-	}
 	now := time.Now()
-	for _, name := range proxies {
-		t.sets.markSeen(now, name)
+	for k, v := range t.tracked {
+		if v.expiry.Before(now) {
+			delete(t.tracked, k)
+		}
 	}
-	count := len(t.sets.proxies)
-	if count < 1 {
-		count = 1
+
+	// degenerate condition: we've just started the tracker or we haven't
+	// successfully connected to any server yet, or network conditions were such
+	// that all tracked proxies have expired - spawn one connection, or we'll
+	// get stuck like that forever
+	if t.inflight == 0 && (len(t.claimed) == 0 || len(t.tracked) == 0) {
+		return true
 	}
-	t.wp.Set(uint64(count))
-}
 
-// Start starts tracking for specified proxy address.
-func (t *Tracker) Start() {
-	t.mu.Lock()
-	defer t.mu.Unlock()
-	t.getOrCreate()
-}
+	desiredGen := make(map[string]int, 8)
+	for _, v := range t.tracked {
+		if v.Generation > desiredGen[v.Group] {
+			desiredGen[v.Group] = v.Generation
+		}
+	}
 
-// Stop stops tracking for specified proxy address.
-func (t *Tracker) Stop() {
-	t.mu.Lock()
-	defer t.mu.Unlock()
-	if t.sets == nil {
-		return
+	desired := make(map[string]struct{}, len(t.tracked))
+	for k, v := range t.tracked {
+		if desiredGen[v.Group] == v.Generation {
+			desired[k] = struct{}{}
+		}
+	}
+
+	// do the set intersection like this because claimed is generally going to
+	// be smaller than desired/tracked
+	desiredClaimed := 0
+	for k := range t.claimed {
+		if _, ok := desired[k]; ok {
+			desiredClaimed++
+		}
+	}
+
+	desiredCount := len(desired)
+	if t.connectionCount > 0 && desiredCount > t.connectionCount {
+		desiredCount = t.connectionCount
 	}
 
-	t.sets = nil
-	t.wp.Set(0)
+	return desiredCount > desiredClaimed+t.inflight
 }
 
-// StopAll permanently deactivates this tracker and cleans
-// up all background goroutines.
-func (t *Tracker) StopAll() {
-	t.cancel()
+// notify signals the run loop that conditions have changed and that the tracker
+// should check again if it should be giving out a lease or not.
+func (t *Tracker) notify() {
+	select {
+	case t.cond <- struct{}{}:
+	default:
+	}
 }
 
-func (t *Tracker) tick() {
+// TrackExpected starts/refreshes tracking for expected proxies.  Called by
+// agents when gossip messages are received.
+func (t *Tracker) TrackExpected(sourceID string, proxies ...Proxy) {
+	if len(proxies) == 0 {
+		return
+	}
+
 	t.mu.Lock()
 	defer t.mu.Unlock()
-	if t.sets == nil {
+	t.notify()
+
+	expiry := time.Now().Add(t.proxyExpiry)
+	for _, p := range proxies {
+		p.expiry = expiry
+		p.deleteAttempt = ""
+		t.tracked[p.Name] = p
+	}
+
+	if sourceID == "" {
```

Comment:
`(*lib/reversetunnel.agent)GetProxyID` is fallible, it can return the empty string if the underlying client is `nil` (maybe it was zeroed out because of a disconnection while we empty out the `handleDiscovery` channel?) or if it can't figure out a server ID from the principals of the remote side - neither of which should ever happen, but if it does I figured that we should not take the gossip info into account for the purpose of deleting proxies, at least.

Reply:
Made irrelevant by the removal of proxy removal by gossip in 95e44718a226abd3eb045f7dd01d0e5e221a2e9e